File: 2020/12/week1/minhyungPark/algorithm/LeetCode1663.kt
/**
 * Leetcode 1663. Smallest String With A Given Numeric Value
 * https://leetcode.com/contest/weekly-contest-216/problems/smallest-string-with-a-given-numeric-value/
 */

class LeetCode1663 {
    fun getSmallestString(n: Int, k: Int): String {
        var remainValue = k
        val stringBuilder = StringBuilder()

        for (i in n downTo 1) {
            val currentMaxValue = remainValue - i + 1
            val numericValue = if (currentMaxValue > 26) 26 else currentMaxValue

            stringBuilder.append(numericValue.toAlphabet())
            remainValue -= numericValue
        }
        return stringBuilder.reverse().toString()
    }

    private fun Int.toAlphabet() = (this - 1 + 'a'.toInt()).toChar()

}

File: 2020/12/week1/1ilsang/algorithm/leetcode1663.js
/**
 * https://leetcode.com/contest/weekly-contest-216/problems/smallest-string-with-a-given-numeric-value/
 * 1663. Smallest String With A Given Numeric Value
 * @param {number} n
 * @param {number} k
 * @return {string}
 */

const getSmallestString = (n, k) => {
  const result = [];
  const offset = "a".charCodeAt() - 1;

  for (let i = n - 1; i >= 0; i--) {
    const num = Math.min(k - i, 26);
    k -= num;
    result.unshift(String.fromCharCode(num + offset));
  }
  return result.join("");
};
